### PR TITLE
fix: enum should be int when enums_as_int = true

### LINKED ISSUE #3162: Swagger docs validate failed when enums_as_ints=true
## 🐛 Bug Report

(A clear and concise description of what the bug is.)

Related issue: https://github.com/go-swagger/go-swagger/issues/2890

## To Reproduce

(Write your steps here:)

Details can be found in this pr https://github.com/hnlq715/proto/pull/3/files

## Expected behavior

Pass validation.
(Write what you thought would happen.)

## Actual Behavior
```
The swagger spec at "gen/go/api/helloworld/v1/helloworld.swagger.json" is invalid against swagger specification 2.0.
See errors below:
- in operation "GreeterService_SayHello", default value in response 200 does not validate its schema
- 200.data.sex.default in body must be of type integer: "string"
- 200.data.sex.default in body should be one of [1 2]
- in operation "GreeterService_SayHello2", default value in response 200 does not validate its schema
- definitions.v1Sex.default in body must be of type integer: "string"
- definitions.v1Sex.default in body should be one of [1 2]
```
(Write what happened. Add screenshots, if applicable.)

## Your Environment

(Environment name, version and operating system.)

### PR DESCRIPTION
**Pull request by hnlq715**
#### References to other Issues or PRs

Fixes #3162

### CODE REVIEW

**Comment by hnlq715** [comment]
With this change, everything works fine.

https://github.com/hnlq715/proto/tree/feat/enums-as-ints

File: protoc-gen-openapiv2/internal/genopenapi/template.go [hunk 1]
```diff
@@ -117,31 +117,52 @@ func listEnumNames(reg *descriptor.Registry, enum *descriptor.Enum) (names []str
 		}
 		names = append(names, value.GetName())
 	}
-	return names
+
+	if len(names) > 0 {
+		return names
+	}
+
+	return nil
```
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/internal/genopenapi/template.go:125]: I don't think this needs to change

File: protoc-gen-openapiv2/internal/genopenapi/template.go [hunk 2]
```diff
@@ -117,31 +117,52 @@ func listEnumNames(reg *descriptor.Registry, enum *descriptor.Enum) (names []str
...
-		numbers = append(numbers, strconv.Itoa(int(value.GetNumber())))
+		numbers = append(numbers, int(value.GetNumber()))
 	}
-	return
+
+	if len(numbers) > 0 {
+		return numbers
+	}
+
+	return nil
```
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/internal/genopenapi/template.go:143]: Please use the same format as the other function
```suggestion
	return numbers
```

File: protoc-gen-openapiv2/internal/genopenapi/template.go [hunk 3]
```diff
@@ -1147,7 +1174,7 @@ func renderServices(services []*descriptor.Service, paths openapiPathsObject, re
 					if parameter.IsRepeated() {
 						core := schemaCore{Type: paramType, Format: paramFormat}
 						if parameter.IsEnum() {
-							var s []string
+							var s interface{}
 							core.Enum = enumNames
 							enumNames = s
```
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/internal/genopenapi/template.go:1179]: I'm confused by this code, but I think we can just do this assignment directly instead
```suggestion
							enumNames = nil
```

File: protoc-gen-openapiv2/internal/genopenapi/template_test.go [hunk 4]
```diff
@@ -8583,6 +8598,10 @@ func TestArrayMessageItemsType(t *testing.T) {
 
 		t.Errorf("applyTemplate(%#v).%s = %v want to be %v", file, name, is, want)
 	}
+	d1, _ := json.Marshal(result.Definitions)
+	d2, _ := json.Marshal(expect)
+	fmt.Println(string(d1))
+	fmt.Println(string(d2))
```
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/internal/genopenapi/template_test.go:8604]: I think these are for debugging? Can we remove them?

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
I think you'll need to regenerate the files, please check https://github.com/grpc-ecosystem/grpc-gateway/blob/main/CONTRIBUTING.md#i-want-to-regenerate-the-files-after-making-changes

File: protoc-gen-openapiv2/internal/genopenapi/template.go (hunk 1 above)
Comment (hnlq715) [reply protoc-gen-openapiv2/internal/genopenapi/template.go:125]: Generated openapi docs would contains `"enum": []` when remove this change, which is not the expected behavior before.

Comment (johanbrandhorst) [reply protoc-gen-openapiv2/internal/genopenapi/template.go:125]: 🤔 that's wild, but okay.

File: protoc-gen-openapiv2/internal/genopenapi/template_test.go (hunk 4 above)
Comment (hnlq715) [reply protoc-gen-openapiv2/internal/genopenapi/template_test.go]: done

File: protoc-gen-openapiv2/internal/genopenapi/template.go (hunk 3 above)
Comment (hnlq715) [reply protoc-gen-openapiv2/internal/genopenapi/template.go:1173]: Nice catch, let's fix this.

**Comment by hnlq715** [comment]
> I think you'll need to regenerate the files, please check https://github.com/grpc-ecosystem/grpc-gateway/blob/main/CONTRIBUTING.md#i-want-to-regenerate-the-files-after-making-changes

I regenerate the files, but nothing changed, seems like there is no use case about `enums_as_ints` in examples, Maybe we should add a case for this?

**Comment by johanbrandhorst** [comment]
> > I think you'll need to regenerate the files, please check https://github.com/grpc-ecosystem/grpc-gateway/blob/main/CONTRIBUTING.md#i-want-to-regenerate-the-files-after-making-changes
> 
> I regenerate the files, but nothing changed, seems like there is no use case about `enums_as_ints` in examples, Maybe we should add a case for this?

That's a great idea, I think you could try picking one of the existing files and adding a new `buf.gen.yaml` file for it.

**Review by spatecon — APPROVED** [review]

**Comment by hnlq715** [comment]
@johanbrandhorst Sorry for the late! 
I have created a new `buf.gen.yaml` for the `enums_as_ints` case in exmaples.

Let me know if anything I can do, thank you in advance :-)

**Review by johanbrandhorst — APPROVED** [review]

**Comment by johanbrandhorst** [comment]
Thanks for your contribution!

**Pull request merged**